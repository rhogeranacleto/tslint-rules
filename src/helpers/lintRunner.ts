import { Configuration, Linter, LintResult } from 'tslint';

interface IMaria {
	src: string;
	// tslint:disable-next-line:no-any
	rule: any;
	fileName?: string;
	fix?: boolean;
}

export function lintRunner(data: IMaria) {

	const { src, rule, fileName = `/tmp/test-${Date.now()}.snap`, fix = false } = data;

	const linter = new Linter({ fix });

	linter.lint(
		fileName,
		src,
		Configuration.parseConfigFile({
			rules: {
				[rule.name || rule]: [true]
			},
			rulesDirectory: 'src'
		})
	);

	return linter.getResult();
}

export function getReplacements(result: LintResult) {

	if (result.fixes) {

		return result.fixes.map(fix => {

			const fixJson = fix.toJson().fix;

			if (fixJson) {

				if (Array.isArray(fixJson)) {

					return fixJson.map(replace => replace.innerText);
				}

				return [fixJson.innerText];
			}
		});
	}
}