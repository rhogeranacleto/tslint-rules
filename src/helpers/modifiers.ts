export function removeLastEmptyLineOfTheBlock(block: string) {

	const lines = block.split('\n');
	const bracket = lines.pop();

	for (let i = lines.length - 1; i > 0; i--) {

		if (lines[i].trim()) {

			break;
		}

		lines.pop();
	}

	lines.push(bracket as string);

	return lines.join('\n');
}