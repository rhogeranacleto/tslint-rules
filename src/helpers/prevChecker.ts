// tslint:disable-next-line:no-implicit-dependencies
import { getPreviousStatement, getPreviousToken } from 'tsutils';
import ts from 'typescript';

export function prevStatementChecker(node: ts.Statement, sourceFile: ts.SourceFile) {

	const prev = getPreviousStatement(node);

	if (prev) {

		return checkPrevLine(node, sourceFile, prev);
	}

	return false;
}

export function prevTokenChecker(node: ts.Node, sourceFile: ts.SourceFile) {

	const prev = getPreviousToken(node);

	if (prev) {

		return checkPrevLine(node, sourceFile, prev);
	}

	return false;
}

function checkPrevLine(node: ts.Node, sourceFile: ts.SourceFile, prev: ts.Node) {

	const startLine = ts.getLineAndCharacterOfPosition(sourceFile, node.getStart()).line;
	const prevLine = ts.getLineAndCharacterOfPosition(sourceFile, prev.getEnd()).line;

	return prevLine === (startLine - 1);
}