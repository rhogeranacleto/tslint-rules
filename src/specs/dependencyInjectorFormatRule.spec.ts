import { lintRunner } from '../helpers/lintRunner';

const RULE = 'dependency-injector-format';

describe('DependecyInjectorFormat tests', () => {

	it.each([
		[`class Den {
			constructor (private readonly on: On) { }
		}`, 0],
		[`class Den {
			constructor (private readonly on: On) {

				console.log(on);
			}
		}`, 0],
		[`class Den {
			constructor (
				private readonly on: On) { }
		}`, 1],
		[`class Den {
			constructor (private readonly on: On) {

			}
		}`, 1],
		[`class Den {
			constructor (private readonly on: On) { console.log(on); }
		}`, 1],
		[`class Dan {
			constructor(adminUser: AdminUser, clientUser?: ClientUser) {
				this.id = adminUser.id;
			}
		}`, 0]
	])('%s', (src, count) => {

		const result = lintRunner({ src, rule: RULE });

		expect(result.errorCount).toBe(count);
	});

	it.each([
		[`class Den {
			constructor (
				private readonly on: On,
				private readonly off: Off) { }
		}`, 0],
		[`class Den {
			constructor (
				private readonly on: On,
				private readonly off: Off
			) { }
		}`, 1],
		[`class Den {
			constructor (private readonly on: On,private readonly off: Off) { }
		}`, 2],
		[`class Den {
			constructor (private readonly on: On,private readonly off: Off) { console.log(on); }
		}`, 3],
		[`class Den {
			constructor (
				private readonly on: On,private readonly off: Off
			) { }
		}`, 2],
		[`class Den {
			constructor (private readonly on: On,
				private readonly off: Off
			) { }
		}`, 3],
		[`class Den {
			constructor (
				private readonly on: On,
				private readonly midle: Midle,
				private readonly off: Off) {
				}
		}`, 1],
		[`class Den {
			constructor (
				private readonly on: On,
				private readonly off: Off
			) {

				console.log(on);
			}
		}`, 0],
		[`class Den {
			constructor (
				private readonly on: On,
				private readonly off: Off) {

				console.log(on);
			}
		}`, 1],
		[`class Den {
			constructor (
				private readonly on: On,
				private readonly off: Off)
			{

				console.log(on);
			}
		}`, 2]
	])('%s', (src, count) => {

		const result = lintRunner({ src, rule: RULE });

		expect(result.errorCount).toBe(count);
	});
});