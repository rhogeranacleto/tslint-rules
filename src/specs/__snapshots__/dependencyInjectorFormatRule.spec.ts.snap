// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DependecyInjectorFormat tests class Den {
			constructor (
				private readonly on: On) { }
		} 1`] = `
"class Den {
			constructor (private readonly on: On) { }"
`;

exports[`DependecyInjectorFormat tests class Den {
			constructor (
				private readonly on: On,
				private readonly off: Off
			) { }
		} 1`] = `
"class Den {
			constructor (
				private readonly on: On,
				private readonly off: Off) { }
		}"
`;

exports[`DependecyInjectorFormat tests class Den {
			constructor (private readonly on: On) {

			}
		} 1`] = `
"class Den {
			constructor (private readonly on: On) {}"
`;

exports[`DependecyInjectorFormat tests class Den {
			constructor (private readonly on: On) { console.log(on); }
		} 1`] = `
"class Den {
			constructor (private readonly on: On) {

 console.log(on); 
}
		}"
`;

exports[`DependecyInjectorFormat tests class Den {
			constructor (private readonly on: On,private readonly off: Off) { }
		} 1`] = `
"class Den {
			constructor (
private readonly on: On,
private readonly off: Off) { } }
		}"
`;
